### ISSUE TITLE
1 Pixel gap between adjacent nodes 

### ISSUE THREAD

**Issue by A-Walrus** [DioxusLabs/taffy#369]
## `taffy` version
0.3.2
## Platform

What platform are you using `taffy` on? Rust
## What you did
```rust
use taffy::prelude::*;

fn main() {
    // First create an instance of Taffy
    let mut taffy = Taffy::new();

    let w = 100.3;

    let w_square = Size {
        width: points(w),
        height: points(w),
    };
    let child_a = taffy
        .new_leaf(Style {
            size: w_square,
            ..Default::default()
        })
        .unwrap();

    let child_b = taffy
        .new_leaf(Style {
            size: w_square,
            ..Default::default()
        })
        .unwrap();

    let root_node = taffy
        .new_with_children(
            Style {
                size: Size {
                    width: points(963.3333),
                    height: points(1000.),
                },
                justify_content: Some(JustifyContent::Center),
                ..Default::default()
            },
            &[child_a, child_b],
        )
        .unwrap();

    taffy.compute_layout(root_node, Size::MAX_CONTENT).unwrap();

    let layout_a = taffy.layout(child_a).unwrap();
    let layout_b = taffy.layout(child_b).unwrap();
    assert_eq!(
        layout_a.location.x + layout_a.size.width,
        layout_b.location.x
    );
}

```
## What went wrong

- what were you expecting?
No gap between adjacent nodes
- what actually happened?
1 Pixel gap
## Additional information

- theories about what might be going wrong
This probably has to do with rounding, see [comment in bevy issue](https://github.com/bevyengine/bevy/issues/7712#issuecomment-1436098176)
- links to related bugs, PRs or discussions
Bug report on Bevy bevyengine/bevy#7712

**Comment by nicoburns** [comment]
Thanks for filing this issue!

A couple of notes:
- This *may* be related to https://github.com/DioxusLabs/taffy/pull/145
- Our gentests are useless here, becuase browsers will happily return naively rounded results which they are clearly not using directly for rendering (there's a 1 pixel gap in the values returned from the APIs, but there visually isn't one in the final rendering).

Fixing this is going to be a little tricky, and is going to require us to come up with our own heuristic to make it work nicely. There isn't really a standard we can follow for this. The other alternative would be just to remove rounding from Taffy entirely, and push this on consumers of Taffy.

**Comment by nicoburns** [comment]
This StackOverflow answer (https://stackoverflow.com/a/5117374/140293) recommends the following algorithm (presented in JavaScript): 
```javascript
var spaceRemaining = list.clientHeight;
var itemsRemaining = items.length;

while (itemsRemaining > 0) {
    var itemHeight = Math.round(spaceRemaining / itemsRemaining);
    items[itemsRemaining - 1].style.height = itemHeight;
    spaceRemaining -= itemHeight;
    itemsRemaining -= 1;
}
```
This only works for evenly sized items, but I think I might be possible to adapt for the more general case. I'm thinking that maybe we want to do something like "round partial pixels down and then distribute remaining available pixels as evenly as possible".

**Comment by nicoburns** [comment]
Ok, some test results (screenshots are from Chrome, but Firefox rendered identically):

### Test Results

#### Test 1 (half pixel)

HTML:
```html
<div id="test-root" style="width:7px;">
  <div style="width:50%;flex-wrap: wrap;">
    <div style="width: 2px;height: 10px;background-color: red"></div>
    <div style="width: 2px;height: 10px;background-color: green"></div>
  </div>
  <div style="width:50%;flex-wrap: wrap;">
    <div style="width: 2px;height: 10px;background-color: yellow"></div>
    <div style="width: 2px;height: 10px;background-color: blue"></div>
  </div>
</div>
```

Zoomed in screenshot with pixel grid:
<img width="257" alt="Chrome partial pixel 0 5 zoomed" src="https://user-images.githubusercontent.com/1007307/220129002-5c367fb5-bf1d-40fb-8410-779953f8c2cf.png">

#### Test 2 (quarter pixel)

HTML:
```html
<div id="test-root" style="width:7px;">
  <div style="width:25%;flex-wrap: wrap;">
    <div style="width: 1px;height: 10px;background-color: red"></div>
    <div style="width: 1px;height: 10px;background-color: green"></div>
  </div>
  <div style="width:25%;flex-wrap: wrap;">
    <div style="width: 1px;height: 10px;background-color: yellow"></div>
    <div style="width: 1px;height: 10px;background-color: blue"></div>
  </div>
  <div style="width:25%;flex-wrap: wrap;">
    <div style="width: 1px;height: 10px;background-color: red"></div>
    <div style="width: 1px;height: 10px;background-color: green"></div>
  </div>
    <div style="width:25%;flex-wrap: wrap;">
    <div style="width: 1px;height: 10px;background-color: yellow"></div>
    <div style="width: 1px;height: 10px;background-color: blue"></div>
  </div>
</div>
```

Zoomed in screenshot with pixel grid:

<img width="229" alt="Chrome partial pixel 0 25 zoomed" src="https://user-images.githubusercontent.com/1007307/220129792-75e5def8-c551-4967-9845-782621047485.png">


### Conclusions

- Fractional pixels do not affect layout at all. As can be clearly be seen in the test 2 screenshot, once the whole-pixel rounding has occured there is actually room to render both grandchildren nodes on the same line for the 1st and 3rd children. But they wrap anyway because the layout is computed using the unrounded float values.
- As shown in test 1, Chrome/Firefox will happily make use of the extra resolution in HiDPI displays to render half-size-logical-pixels perfectly.
- When exceeding the available display resolution (as in test 2), elements will be selectively rounded up/down in size in order to make the overall sizes fit perfectly. These rounded sizes are used for painting.

**Comment by nicoburns** [comment]
Looks like Yoga had exactly the same problem before us (See: https://github.com/facebook/yoga/commit/aa5b296ac78f7a22e1aeaf4891243c6bb76488e2). We actually had the fix, but it was broken by #145 because our gentests weren't configured to round in the same way.

**Comment by ickshonpe** [comment]
I'll update bevy with this and see how it looks.
